File: apps/api/src/lib/state-machine.ts
import { logger } from '../utils/logger';

/**
 * Vocabotics Project State Machine
 * Orchestrates the workflow from vision → production
 */

export enum ProjectState {
  // Phase 1: Vision & Requirements
  VISION_INPUT = 'vision_input',
  PRD_GENERATION = 'prd_generation',
  PRD_REVIEW = 'prd_review',

  // Phase 2: Architecture & Design
  ARCHITECTURE_DESIGN = 'architecture_design',
  ARCHITECTURE_REVIEW = 'architecture_review',
  SCHEMA_GENERATION = 'schema_generation',
  API_SPECIFICATION = 'api_specification',

  // Phase 3: Implementation
  CODE_GENERATION = 'code_generation',
  CODE_REVIEW = 'code_review',
  TEST_GENERATION = 'test_generation',

  // Phase 4: Testing & Quality
  INTEGRATION_TESTING = 'integration_testing',
  VISUAL_TESTING = 'visual_testing',
  QUALITY_REVIEW = 'quality_review',

  // Phase 5: Deployment
  DEPLOYMENT_PREP = 'deployment_prep',
  DEPLOYMENT = 'deployment',
  DEPLOYED = 'deployed',

  // Error states
  ERROR = 'error',
  PAUSED = 'paused',
}

export enum TransitionTrigger {
  // User actions
  USER_SUBMIT = 'user_submit',
  USER_APPROVE = 'user_approve',
  USER_REJECT = 'user_reject',
  USER_PAUSE = 'user_pause',
  USER_RESUME = 'user_resume',

  // System actions
  AI_COMPLETE = 'ai_complete',
  AI_ERROR = 'ai_error',
  TEST_PASS = 'test_pass',
  TEST_FAIL = 'test_fail',
  VALIDATION_PASS = 'validation_pass',
  VALIDATION_FAIL = 'validation_fail',

  // Integration actions
  GITHUB_COMMIT = 'github_commit',
  DEPLOYMENT_SUCCESS = 'deployment_success',
  DEPLOYMENT_FAILURE = 'deployment_failure',
}

export interface StateTransition {
  from: ProjectState;
  to: ProjectState;
  trigger: TransitionTrigger;
  condition?: (context: any) => boolean;
  action?: (context: any) => Promise<void>;
}

export interface StateMachineContext {
  projectId: string;
  currentState: ProjectState;
  metadata: Record<string, any>;
  history: {
    state: ProjectState;
    timestamp: Date;
    triggeredBy: string;
  }[];
}

export class ProjectStateMachine {
  private transitions: StateTransition[] = [
    // Phase 1: Vision → PRD
    {
      from: ProjectState.VISION_INPUT,
      to: ProjectState.PRD_GENERATION,
      trigger: TransitionTrigger.USER_SUBMIT,
    },
    {
      from: ProjectState.PRD_GENERATION,
      to: ProjectState.PRD_REVIEW,
      trigger: TransitionTrigger.AI_COMPLETE,
    },
    {
      from: ProjectState.PRD_GENERATION,
      to: ProjectState.ERROR,
      trigger: TransitionTrigger.AI_ERROR,
    },
    {
      from: ProjectState.PRD_REVIEW,
      to: ProjectState.ARCHITECTURE_DESIGN,
      trigger: TransitionTrigger.USER_APPROVE,
    },
    {
      from: ProjectState.PRD_REVIEW,
      to: ProjectState.PRD_GENERATION,
      trigger: TransitionTrigger.USER_REJECT,
    },

    // Phase 2: Architecture → Schema → API
    {
      from: ProjectState.ARCHITECTURE_DESIGN,
      to: ProjectState.ARCHITECTURE_REVIEW,
      trigger: TransitionTrigger.AI_COMPLETE,
    },
    {
      from: ProjectState.ARCHITECTURE_DESIGN,
      to: ProjectState.ERROR,
      trigger: TransitionTrigger.AI_ERROR,
    },
    {
      from: ProjectState.ARCHITECTURE_REVIEW,
      to: ProjectState.SCHEMA_GENERATION,
      trigger: TransitionTrigger.USER_APPROVE,
    },
    {
      from: ProjectState.ARCHITECTURE_REVIEW,
      to: ProjectState.ARCHITECTURE_DESIGN,
      trigger: TransitionTrigger.USER_REJECT,
    },
    {
      from: ProjectState.SCHEMA_GENERATION,
      to: ProjectState.API_SPECIFICATION,
      trigger: TransitionTrigger.AI_COMPLETE,
    },
    {
      from: ProjectState.API_SPECIFICATION,
      to: ProjectState.CODE_GENERATION,
      trigger: TransitionTrigger.AI_COMPLETE,
    },

    // Phase 3: Code Generation → Testing
    {
      from: ProjectState.CODE_GENERATION,
      to: ProjectState.CODE_REVIEW,
      trigger: TransitionTrigger.AI_COMPLETE,
    },
    {
      from: ProjectState.CODE_GENERATION,
      to: ProjectState.ERROR,
      trigger: TransitionTrigger.AI_ERROR,
    },
    {
      from: ProjectState.CODE_REVIEW,
      to: ProjectState.TEST_GENERATION,
      trigger: TransitionTrigger.USER_APPROVE,
    },
    {
      from: ProjectState.CODE_REVIEW,
      to: ProjectState.CODE_GENERATION,
      trigger: TransitionTrigger.USER_REJECT,
    },
    {
      from: ProjectState.TEST_GENERATION,
      to: ProjectState.INTEGRATION_TESTING,
      trigger: TransitionTrigger.AI_COMPLETE,
    },

    // Phase 4: Testing & Quality
    {
      from: ProjectState.INTEGRATION_TESTING,
      to: ProjectState.VISUAL_TESTING,
      trigger: TransitionTrigger.TEST_PASS,
    },
    {
      from: ProjectState.INTEGRATION_TESTING,
      to: ProjectState.CODE_GENERATION,
      trigger: TransitionTrigger.TEST_FAIL,
    },
    {
      from: ProjectState.VISUAL_TESTING,
      to: ProjectState.QUALITY_REVIEW,
      trigger: TransitionTrigger.TEST_PASS,
    },
    {
      from: ProjectState.VISUAL_TESTING,
      to: ProjectState.CODE_GENERATION,
      trigger: TransitionTrigger.TEST_FAIL,
    },
    {
      from: ProjectState.QUALITY_REVIEW,
      to: ProjectState.DEPLOYMENT_PREP,
      trigger: TransitionTrigger.USER_APPROVE,
    },
    {
      from: ProjectState.QUALITY_REVIEW,
      to: ProjectState.CODE_GENERATION,
      trigger: TransitionTrigger.USER_REJECT,
    },

    // Phase 5: Deployment
    {
      from: ProjectState.DEPLOYMENT_PREP,
      to: ProjectState.DEPLOYMENT,
      trigger: TransitionTrigger.VALIDATION_PASS,
    },
    {
      from: ProjectState.DEPLOYMENT,
      to: ProjectState.DEPLOYED,
      trigger: TransitionTrigger.DEPLOYMENT_SUCCESS,
    },
    {
      from: ProjectState.DEPLOYMENT,
      to: ProjectState.ERROR,
      trigger: TransitionTrigger.DEPLOYMENT_FAILURE,
    },

    // Pause/Resume from any state
    {
      from: ProjectState.PRD_GENERATION,
      to: ProjectState.PAUSED,
      trigger: TransitionTrigger.USER_PAUSE,
    },
    {
      from: ProjectState.ARCHITECTURE_DESIGN,
      to: ProjectState.PAUSED,
      trigger: TransitionTrigger.USER_PAUSE,
    },
    {
      from: ProjectState.CODE_GENERATION,
      to: ProjectState.PAUSED,
      trigger: TransitionTrigger.USER_PAUSE,
    },
  ];

  /**
   * Check if transition is valid
   */
  canTransition(
    from: ProjectState,
    trigger: TransitionTrigger,
    context?: any
  ): boolean {
    const transition = this.transitions.find(
      (t) => t.from === from && t.trigger === trigger
    );

    if (!transition) {
      return false;
    }

    if (transition.condition && context) {
      return transition.condition(context);
    }

    return true;
  }

  /**
   * Get next state for a given trigger
   */
  getNextState(
    from: ProjectState,
    trigger: TransitionTrigger
  ): ProjectState | null {
    const transition = this.transitions.find(
      (t) => t.from === from && t.trigger === trigger
    );

    return transition ? transition.to : null;
  }

  /**
   * Execute transition with action
   */
  async transition(
    context: StateMachineContext,
    trigger: TransitionTrigger,
    triggeredBy: string
  ): Promise<ProjectState> {
    const { currentState } = context;

    if (!this.canTransition(currentState, trigger, context)) {
      logger.error('Invalid state transition', {
        projectId: context.projectId,
        from: currentState,
        trigger,
      });
      throw new Error(
        `Invalid transition from ${currentState} with trigger ${trigger}`
      );
    }

    const transition = this.transitions.find(
      (t) => t.from === currentState && t.trigger === trigger
    );

    if (!transition) {
      throw new Error('Transition not found');
    }

    const nextState = transition.to;

    // Execute transition action if defined
    if (transition.action) {
      try {
        await transition.action(context);
      } catch (error) {
        logger.error('Transition action failed', {
          projectId: context.projectId,
          from: currentState,
          to: nextState,
          error: error instanceof Error ? error.message : 'Unknown error',
        });
        throw error;
      }
    }

    // Update context
    context.currentState = nextState;
    context.history.push({
      state: nextState,
      timestamp: new Date(),
      triggeredBy,
    });

    logger.info('State transition executed', {
      projectId: context.projectId,
      from: currentState,
      to: nextState,
      trigger,
      triggeredBy,
    });

    return nextState;
  }

  /**
   * Get all possible transitions from current state
   */
  getPossibleTransitions(state: ProjectState): {
    trigger: TransitionTrigger;
    nextState: ProjectState;
  }[] {
    return this.transitions
      .filter((t) => t.from === state)
      .map((t) => ({
        trigger: t.trigger,
        nextState: t.to,
      }));
  }

  /**
   * Get state phase
   */
  getPhase(state: ProjectState): string {
    const phases: Record<string, ProjectState[]> = {
      'Vision & Requirements': [
        ProjectState.VISION_INPUT,
        ProjectState.PRD_GENERATION,
        ProjectState.PRD_REVIEW,
      ],
      'Architecture & Design': [
        ProjectState.ARCHITECTURE_DESIGN,
        ProjectState.ARCHITECTURE_REVIEW,
        ProjectState.SCHEMA_GENERATION,
        ProjectState.API_SPECIFICATION,
      ],
      Implementation: [
        ProjectState.CODE_GENERATION,
        ProjectState.CODE_REVIEW,
        ProjectState.TEST_GENERATION,
      ],
      'Testing & Quality': [
        ProjectState.INTEGRATION_TESTING,
        ProjectState.VISUAL_TESTING,
        ProjectState.QUALITY_REVIEW,
      ],
      Deployment: [
        ProjectState.DEPLOYMENT_PREP,
        ProjectState.DEPLOYMENT,
        ProjectState.DEPLOYED,
      ],
    };

    for (const [phase, states] of Object.entries(phases)) {
      if (states.includes(state)) {
        return phase;
      }
    }

    return 'Other';
  }

  /**
   * Calculate progress percentage
   */
  getProgress(state: ProjectState): number {
    const stateOrder = [
      ProjectState.VISION_INPUT, // 0%
      ProjectState.PRD_GENERATION, // 5%
      ProjectState.PRD_REVIEW, // 10%
      ProjectState.ARCHITECTURE_DESIGN, // 20%
      ProjectState.ARCHITECTURE_REVIEW, // 25%
      ProjectState.SCHEMA_GENERATION, // 30%
      ProjectState.API_SPECIFICATION, // 35%
      ProjectState.CODE_GENERATION, // 50%
      ProjectState.CODE_REVIEW, // 60%
      ProjectState.TEST_GENERATION, // 65%
      ProjectState.INTEGRATION_TESTING, // 75%
      ProjectState.VISUAL_TESTING, // 80%
      ProjectState.QUALITY_REVIEW, // 85%
      ProjectState.DEPLOYMENT_PREP, // 90%
      ProjectState.DEPLOYMENT, // 95%
      ProjectState.DEPLOYED, // 100%
    ];

    const index = stateOrder.indexOf(state);
    if (index === -1) return 0;

    return Math.round((index / (stateOrder.length - 1)) * 100);
  }
}

// Export singleton instance
export const stateMachine = new ProjectStateMachine();

File: vocabotics-cli/vocabotics/state_machine.py
"""
Vocabotics Project State Machine
Orchestrates the workflow from vision → production
"""

from enum import Enum
from typing import Dict, List, Optional, Any
from datetime import datetime
import json


class ProjectState(Enum):
    """Project workflow states"""
    # Phase 1: Vision & Requirements
    VISION_INPUT = "vision_input"
    PRD_GENERATION = "prd_generation"
    PRD_REVIEW = "prd_review"

    # Phase 2: Architecture & Design
    ARCHITECTURE_DESIGN = "architecture_design"
    ARCHITECTURE_REVIEW = "architecture_review"
    SCHEMA_GENERATION = "schema_generation"
    API_SPECIFICATION = "api_specification"

    # Phase 3: Implementation
    CODE_GENERATION = "code_generation"
    CODE_REVIEW = "code_review"
    TEST_GENERATION = "test_generation"

    # Phase 4: Testing & Quality
    INTEGRATION_TESTING = "integration_testing"
    VISUAL_TESTING = "visual_testing"
    QUALITY_REVIEW = "quality_review"

    # Phase 5: Deployment
    DEPLOYMENT_PREP = "deployment_prep"
    DEPLOYMENT = "deployment"
    DEPLOYED = "deployed"

    # Error states
    ERROR = "error"
    PAUSED = "paused"


class TransitionTrigger(Enum):
    """State transition triggers"""
    # User actions
    USER_SUBMIT = "user_submit"
    USER_APPROVE = "user_approve"
    USER_REJECT = "user_reject"
    USER_PAUSE = "user_pause"
    USER_RESUME = "user_resume"

    # System actions
    AI_COMPLETE = "ai_complete"
    AI_ERROR = "ai_error"
    TEST_PASS = "test_pass"
    TEST_FAIL = "test_fail"
    VALIDATION_PASS = "validation_pass"
    VALIDATION_FAIL = "validation_fail"

    # Integration actions
    GITHUB_COMMIT = "github_commit"
    DEPLOYMENT_SUCCESS = "deployment_success"
    DEPLOYMENT_FAILURE = "deployment_failure"


class StateTransition:
    """State transition definition"""
    def __init__(self, from_state: ProjectState, to_state: ProjectState,
                 trigger: TransitionTrigger):
        self.from_state = from_state
        self.to_state = to_state
        self.trigger = trigger


class StateMachineContext:
    """State machine context"""
    def __init__(self, project_id: str, current_state: ProjectState):
        self.project_id = project_id
        self.current_state = current_state
        self.metadata: Dict[str, Any] = {}
        self.history: List[Dict[str, Any]] = []

    def to_dict(self) -> Dict:
        """Serialize to dictionary"""
        return {
            "project_id": self.project_id,
            "current_state": self.current_state.value,
            "metadata": self.metadata,
            "history": self.history
        }

    @classmethod
    def from_dict(cls, data: Dict) -> 'StateMachineContext':
        """Deserialize from dictionary"""
        context = cls(
            data["project_id"],
            ProjectState(data["current_state"])
        )
        context.metadata = data.get("metadata", {})
        context.history = data.get("history", [])
        return context


class ProjectStateMachine:
    """State machine for project workflow"""

    def __init__(self):
        self.transitions = self._init_transitions()

    def _init_transitions(self) -> List[StateTransition]:
        """Initialize all valid transitions"""
        return [
            # Phase 1: Vision → PRD
            StateTransition(ProjectState.VISION_INPUT, ProjectState.PRD_GENERATION,
                          TransitionTrigger.USER_SUBMIT),
            StateTransition(ProjectState.PRD_GENERATION, ProjectState.PRD_REVIEW,
                          TransitionTrigger.AI_COMPLETE),
            StateTransition(ProjectState.PRD_GENERATION, ProjectState.ERROR,
                          TransitionTrigger.AI_ERROR),
            StateTransition(ProjectState.PRD_REVIEW, ProjectState.ARCHITECTURE_DESIGN,
                          TransitionTrigger.USER_APPROVE),
            StateTransition(ProjectState.PRD_REVIEW, ProjectState.PRD_GENERATION,
                          TransitionTrigger.USER_REJECT),

            # Phase 2: Architecture → Schema → API
            StateTransition(ProjectState.ARCHITECTURE_DESIGN, ProjectState.ARCHITECTURE_REVIEW,
                          TransitionTrigger.AI_COMPLETE),
            StateTransition(ProjectState.ARCHITECTURE_DESIGN, ProjectState.ERROR,
                          TransitionTrigger.AI_ERROR),
            StateTransition(ProjectState.ARCHITECTURE_REVIEW, ProjectState.SCHEMA_GENERATION,
                          TransitionTrigger.USER_APPROVE),
            StateTransition(ProjectState.ARCHITECTURE_REVIEW, ProjectState.ARCHITECTURE_DESIGN,
                          TransitionTrigger.USER_REJECT),
            StateTransition(ProjectState.SCHEMA_GENERATION, ProjectState.API_SPECIFICATION,
                          TransitionTrigger.AI_COMPLETE),
            StateTransition(ProjectState.API_SPECIFICATION, ProjectState.CODE_GENERATION,
                          TransitionTrigger.AI_COMPLETE),

            # Phase 3: Code Generation → Testing
            StateTransition(ProjectState.CODE_GENERATION, ProjectState.CODE_REVIEW,
                          TransitionTrigger.AI_COMPLETE),
            StateTransition(ProjectState.CODE_GENERATION, ProjectState.ERROR,
                          TransitionTrigger.AI_ERROR),
            StateTransition(ProjectState.CODE_REVIEW, ProjectState.TEST_GENERATION,
                          TransitionTrigger.USER_APPROVE),
            StateTransition(ProjectState.CODE_REVIEW, ProjectState.CODE_GENERATION,
                          TransitionTrigger.USER_REJECT),
            StateTransition(ProjectState.TEST_GENERATION, ProjectState.INTEGRATION_TESTING,
                          TransitionTrigger.AI_COMPLETE),

            # Phase 4: Testing & Quality
            StateTransition(ProjectState.INTEGRATION_TESTING, ProjectState.VISUAL_TESTING,
                          TransitionTrigger.TEST_PASS),
            StateTransition(ProjectState.INTEGRATION_TESTING, ProjectState.CODE_GENERATION,
                          TransitionTrigger.TEST_FAIL),
            StateTransition(ProjectState.VISUAL_TESTING, ProjectState.QUALITY_REVIEW,
                          TransitionTrigger.TEST_PASS),
            StateTransition(ProjectState.VISUAL_TESTING, ProjectState.CODE_GENERATION,
                          TransitionTrigger.TEST_FAIL),
            StateTransition(ProjectState.QUALITY_REVIEW, ProjectState.DEPLOYMENT_PREP,
                          TransitionTrigger.USER_APPROVE),
            StateTransition(ProjectState.QUALITY_REVIEW, ProjectState.CODE_GENERATION,
                          TransitionTrigger.USER_REJECT),

            # Phase 5: Deployment
            StateTransition(ProjectState.DEPLOYMENT_PREP, ProjectState.DEPLOYMENT,
                          TransitionTrigger.VALIDATION_PASS),
            StateTransition(ProjectState.DEPLOYMENT, ProjectState.DEPLOYED,
                          TransitionTrigger.DEPLOYMENT_SUCCESS),
            StateTransition(ProjectState.DEPLOYMENT, ProjectState.ERROR,
                          TransitionTrigger.DEPLOYMENT_FAILURE),
        ]

    def can_transition(self, from_state: ProjectState, trigger: TransitionTrigger) -> bool:
        """Check if transition is valid"""
        return any(
            t.from_state == from_state and t.trigger == trigger
            for t in self.transitions
        )

    def get_next_state(self, from_state: ProjectState,
                      trigger: TransitionTrigger) -> Optional[ProjectState]:
        """Get next state for a given trigger"""
        for t in self.transitions:
            if t.from_state == from_state and t.trigger == trigger:
                return t.to_state
        return None

    def transition(self, context: StateMachineContext,
                  trigger: TransitionTrigger, triggered_by: str) -> ProjectState:
        """Execute state transition"""
        current_state = context.current_state

        if not self.can_transition(current_state, trigger):
            raise ValueError(
                f"Invalid transition from {current_state.value} with trigger {trigger.value}"
            )

        next_state = self.get_next_state(current_state, trigger)
        if not next_state:
            raise ValueError("Transition not found")

        # Update context
        context.current_state = next_state
        context.history.append({
            "state": next_state.value,
            "timestamp": datetime.now().isoformat(),
            "triggered_by": triggered_by
        })

        return next_state

    def get_phase(self, state: ProjectState) -> str:
        """Get the phase name for a state"""
        phases = {
            "Vision & Requirements": [
                ProjectState.VISION_INPUT,
                ProjectState.PRD_GENERATION,
                ProjectState.PRD_REVIEW,
            ],
            "Architecture & Design": [
                ProjectState.ARCHITECTURE_DESIGN,
                ProjectState.ARCHITECTURE_REVIEW,
                ProjectState.SCHEMA_GENERATION,
                ProjectState.API_SPECIFICATION,
            ],
            "Implementation": [
                ProjectState.CODE_GENERATION,
                ProjectState.CODE_REVIEW,
                ProjectState.TEST_GENERATION,
            ],
            "Testing & Quality": [
                ProjectState.INTEGRATION_TESTING,
                ProjectState.VISUAL_TESTING,
                ProjectState.QUALITY_REVIEW,
            ],
            "Deployment": [
                ProjectState.DEPLOYMENT_PREP,
                ProjectState.DEPLOYMENT,
                ProjectState.DEPLOYED,
            ],
        }

        for phase, states in phases.items():
            if state in states:
                return phase
        return "Other"

    def get_progress(self, state: ProjectState) -> int:
        """Calculate progress percentage"""
        state_order = [
            ProjectState.VISION_INPUT,           # 0%
            ProjectState.PRD_GENERATION,         # 5%
            ProjectState.PRD_REVIEW,             # 10%
            ProjectState.ARCHITECTURE_DESIGN,    # 20%
            ProjectState.ARCHITECTURE_REVIEW,    # 25%
            ProjectState.SCHEMA_GENERATION,      # 30%
            ProjectState.API_SPECIFICATION,      # 35%
            ProjectState.CODE_GENERATION,        # 50%
            ProjectState.CODE_REVIEW,            # 60%
            ProjectState.TEST_GENERATION,        # 65%
            ProjectState.INTEGRATION_TESTING,    # 75%
            ProjectState.VISUAL_TESTING,         # 80%
            ProjectState.QUALITY_REVIEW,         # 85%
            ProjectState.DEPLOYMENT_PREP,        # 90%
            ProjectState.DEPLOYMENT,             # 95%
            ProjectState.DEPLOYED,               # 100%
        ]

        try:
            index = state_order.index(state)
            return round((index / (len(state_order) - 1)) * 100)
        except ValueError:
            return 0
